### PR TITLE
feat(snowflake): add pure-Python Snowflake SQL API adapter

### PR DESCRIPTION
**Pull request by hampsterx**
## Summary

- Adds a `[snowflake]` extra and `create_snowflake_client` / `create_async_snowflake_client` so Lambdas can query Snowflake without the 75MB `snowflake-connector-python`. The adapter wraps the pure-Python `snowflake-sql-api` client (httpx), which cold-starts in ~1.2s vs the connector's 8-20s.
- Keypair credentials load from AWS Secrets Manager by default, with environment-variable and explicit-argument overrides. Resolution is field-by-field with strict `explicit > env > Secrets Manager` precedence and source-aware passphrase pairing (a key is only ever paired with a passphrase from its own source).
- Applies NUI session defaults (`TIMEZONE=Pacific/Auckland`, role `NUI_LAMBDA`), both overridable, and wires a redacting query-logging hook that logs the statement text plus bind-parameter count, never the bind values, JWTs, key material, or the passphrase.
- Sync client is the default; an async factory mirrors it for Lambdas already on an event loop.
- The dependency is a pre-release git-SHA pin of the public `snowflake-sql-api` repo, kept out of `[all]` so nothing pulls the git dependency implicitly. It will be repinned to a PyPI release once that package ships.

`WEEK_START` is intentionally not a session default: the Snowflake SQL API rejects it in the request `parameters` field (HTTP 400, verified live) and is stateless per statement, so it cannot be applied per call. Set it on the Snowflake user/role default if a consumer needs Monday-first weeks.

## Backwards compatibility

Additive only: a new optional extra and a new module. No behavior change for existing consumers; nothing installs the extra implicitly. The `nui_lambda_shared_utils` deprecation shim re-exports the new surface.

## Review

Internal: Pass 1 (bugs) clean, Pass 2 (polish) 5 items auto-fixed. External (cross-family Codex + Kimi, two rounds): no major defects; an explicit-key-vs-env precedence bug and an empty-key edge case were caught and fixed with regression tests; both reviewers confirmed credential precedence, redaction, sync/async parity, kwargs assembly, the missing-extra guard, and lazy-export wiring are correct.

## Changes

```
 AGENTS.md                                   |   1 +
 README.md                                   |  19 +
 nui_lambda_shared_utils/snowflake_client.py |   3 +
 nui_shared_utils/__init__.py                |  12 +
 nui_shared_utils/snowflake_client.py        | 422 +++++++++++++++++++++
 pyproject.toml                              |   6 +
 requirements-test.txt                       |   5 +-
 setup.py                                    |   7 +
 tests/test_snowflake_client.py              | 559 ++++++++++++++++++++++++++++
 9 files changed, 1033 insertions(+), 1 deletion(-)
```

## Test plan

- [x] 36 adapter unit tests, 100% coverage on the new module (moto for Secrets Manager)
- [x] Full package suite green (555 passed)
- [x] black + mypy clean on the new module; `python -m build` + `twine check` pass with the git-pin extra
- [x] Live smoke against dev Snowflake through the adapter: TIMEZONE default applied and overridable, multi-row fetch, server-side bind round-trip, redaction confirmed
- [ ] CI matrix green on Python 3.9-3.12

## Summary by CodeRabbit

* **New Features**
  * Added a Snowflake SQL API integration with synchronous and asynchronous client factories.
  * Implemented credential resolution with clear precedence across explicit inputs, environment variables, and AWS Secrets Manager.
  * Added a redacting query-logging hook that truncates SQL and avoids bind value/secret leakage.
* **Documentation**
  * Updated installation extras and added Snowflake (SQL API) usage examples.
* **Tests**
  * Added an offline test suite covering credential precedence, validation, redaction behavior, sync/async parity, and missing-extra error handling.

### CODE REVIEW

**Comment by hampsterx** [comment]
@coderabbitai review

**Comment by hampsterx** [comment]
Applied the README nitpick: the Snowflake example now overrides `timezone=` and `role=` to show the NUI defaults are configurable for any account.

File: README.md [hunk 1]
```diff
@@ -198,6 +199,31 @@ async with db.get_connection() as conn:
...
+    warehouse="COMPUTE_WH",
+    database="ANALYTICS",
+    timezone="UTC",          # override the Pacific/Auckland default
+    role="MY_APP_ROLE",      # override the NUI_LAMBDA default
+)
+rows = client.query("SELECT id, name FROM orders WHERE status = ?", ["confirmed"])
+```
+
+Sync is the default; use `create_async_snowflake_client(...)` inside a Lambda
+already on an event loop. Requires the `[snowflake]` extra.
```
[the remark this answers is not in the document]
Comment (hampsterx) [reply README.md:225]: Skipped, markdownlint-cli2 with the repo config reports 0 errors for README.md (verified locally). The MD031 hits are in docs/guides/shared-types.md, which is pre-existing and out of scope for this PR.

File: nui_shared_utils/snowflake_client.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+
+def get_snowflake_credentials(
+    secret_name: Optional[str] = None,
+    *,
+    account: Optional[str] = None,
+    user: Optional[str] = None,
+    private_key: Optional[bytes] = None,
+    private_key_passphrase: Optional[str] = None,
+) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (hampsterx) [reply nui_shared_utils/snowflake_client.py:139]: Fixed, widened `private_key` to `Optional[Union[bytes, str]]` across `get_snowflake_credentials` and both factories; `_as_key_bytes` already accepted PEM str (covered by test_explicit_string_key_coerced_to_bytes).

**Comment by hampsterx** [comment]
Addressed the review:
- **README `[all]` wording**: clarified to `# Core optional integrations (excludes Snowflake)` (Snowflake is intentionally out of `[all]`).
- **pytest markers**: added module-level `pytestmark = pytest.mark.unit` (repo convention; these tests are all offline/fast).
- **`private_key` type**: widened to `Optional[Union[bytes, str]]`.
- **README MD031**: not reproduced (markdownlint-cli2 reports README.md clean; flagged file is a pre-existing doc, out of scope).
